ulib/FStar.List.Tot.Properties.fst

```diff
@@ -285,6 +302,101 @@ let rec map_lemma f l =
     | [] -> ()
     | h::t -> map_lemma f t
 
+(** Properties about unsnoc *)
+
+(** [unsnoc] is the inverse of [snoc] *)
+val lemma_unsnoc_snoc: #a:Type -> l:list a{length l > 0} ->
+  Lemma (requires True)
+    (ensures (let l', x = unsnoc l in snoc l' x == l))
+    [SMTPat (unsnoc l)]
+let lemma_unsnoc_snoc #a l =
+  let l', x = unsnoc l in
+  let l1, l2 = l', [x] in
+  lemma_splitAt_snd_length (length l - 1) l;
+  // assert ((l1, l2) == splitAt (length l - 1) l);
+  let rec aux (l:list a{length l > 0}) :
+    Lemma (let l1, l2 = splitAt (length l - 1) l in
+           append l1 l2 == l) =
+    if length l = 1 then () else aux (tl l) in
+  aux l
+
+(** [unsnoc] followed by [append] can be connected to the same vice-versa. *)
+let rec lemma_unsnoc_append (#a:Type) (l1 l2:list a) :
+  Lemma
+    (requires (length l2 > 0)) // the [length l2 = 0] is trivial
+    (ensures (
+        let as, a = unsnoc (l1 @ l2) in
+        let bs, b = unsnoc l2 in
+        as == l1 @ bs /\ a == b)) =
+  match l1 with
+  | [] -> ()
+  | _ :: l1' -> lemma_unsnoc_append l1' l2
+
+(** [unsnoc] gives you [last] element, which is [index]ed at [length l - 1] *)
+let rec lemma_unsnoc_is_last (#t:Type) (l:list t) :
+  Lemma
+    (requires (length l > 0))
+    (ensures (snd (unsnoc l) == last l /\ snd (unsnoc l) == index l (length l - 1))) =
+  match l with
+  | [_] -> ()
+  | _ -> lemma_unsnoc_is_last (tl l)
+
+(** [index]ing on the left part of an [unsnoc]d list is the same as indexing
+    the original list. *)
+let rec lemma_unsnoc_index (#t:Type) (l:list t) (i:nat) :
+  Lemma
+    (requires (length l > 0 /\ i < length l - 1))
+    (ensures (
+        index (fst (unsnoc l)) i == index l i)) =
+  match i with
+  | 0 -> ()
+  | _ -> lemma_unsnoc_index (tl l) (i - 1)
+
+(** Definition and properties about [split_using] *)
+
+(** [split_using] splits a list at the first instance of finding an
+    element in it. *)
+let rec split_using (#t:Type) (l:list t) (x:t{x `memP` l}) :
+  GTot (r:(list t * list t)) =
```

Comment:
In what cases do you need to split a list like this when `t` is not an `eqtype`?

Reply:
In my dll proof, I hold on to an `erased (list (pointer (node t)))`, and need to regularly split this given only a `pointer (node t)`. Since this is an `erased`, it is sufficient for me to be able to do this in the `GTot` world.

I'd be glad to hear alternative suggestions for this, but I believe it to be a useful thing to be able to do at the ghost-level.

Reply:
I can see how that might be useful.

I'm a bit sad to see StrongExcludedMiddle in the middle of the list library. But, I don't have an alternative to offer you without an in depth review and potential overhaul of your proof of a  doubly linked list. At a distance, if your `erased (list (pointer (node t)))` represents the ghost state of your linked list, then why does it contain pointers to nodes rather than the nodes themselves. Anyway, that's just a shot in the dark not knowing much about your proof.

At this stage, I am okay with accepting this function but it would be good to document it prominently saying that it uses this axiom internally.

Reply:
Yes, it does indeed represent the ghost state of the linked list. I had moved between `erased (list (pointer (node t)))` and `erased (list (node t))` multiple times in the process of building the proof, and from experience, I realized that holding on to the list of pointers helps to tremendously simplify the proof since a bunch of properties that are essential (almost counterintuitively) can be talked about without refering to heaps. By simplification, I mean both "understandability" of the proof, as well as the time the proof takes to verify successfully.

For now, I have added in a prominent `NOTE` in the doc string for this.